### PR TITLE
Intruduce portforward lib into test-framework

### LINKED ISSUE #6173: [test-framework] Ensure that the port-forwarder lifecycle is limited.
The current port-forwarding logic in the test-framework launches kubectl as a process to forward ports. This has potential cleanup issues: if the user terminates the go test, or if there is a panic, then the clean up would not be executed and the port forwarder processes will keep running.

There are two potential solutions to this problem:

- Use the port-forwarder as a library (https://github.com/kubernetes/client-go/blob/master/tools/portforward/portforward.go)
- Use a wrapper around kubectl that checks for a heartbeat (e.g. a shell program that launches port-forwarder in the background and checks for timestamp/process list updates to detect if the test is running).

### PR DESCRIPTION
**Pull request by lichuqiang**
Fix #6173 
Replace "kubectl port-forward" with that in client-go,
to make sure  the port-forwarder lifecycle is limited

### CODE REVIEW

File: pkg/test/framework/components/policybackend/policybackend.go [hunk 1]
```diff
@@ -197,9 +198,22 @@ func (c *kubeComponent) Init(ctx environment.ComponentContext, deps map[dependen
 	addressInCluster := fmt.Sprintf("%s:%d", svc.Spec.ClusterIP, svc.Spec.Ports[0].TargetPort.IntVal)
 	scope.Debugf("Policy Backend in-cluster address: %s", addressInCluster)
 
-	forwarder := kube.NewPortForwarder(ctx.Settings().KubeConfig, pod.Namespace, pod.Name, int(svc.Spec.Ports[0].TargetPort.IntVal))
-	if err = forwarder.Start(); err != nil {
-		return nil, err
+	rc, config, err := kube.NewRestClient(ctx.Settings().KubeConfig, "")
```
Comment (nmittler) [inline pkg/test/framework/components/policybackend/policybackend.go]: There's a lot of copy & paste here .... could we make a utility to help here?

File: pkg/test/kube/forwarder.go [hunk 2]
```diff
@@ -1,107 +1,184 @@
...
 
-// Address is the local address for the port forwarder.
-func (p *PortForwarder) Address() string {
-	return p.address
+func (f *defaultPortForwarder) Address() string {
+	return f.address
+}
+
+func (f *defaultPortForwarder) Close() {
+	f.forwarder.Close()
```
Comment (nmittler) [inline pkg/test/kube/forwarder.go]: Does this send something to the `stopCh`? If so, maybe we keep this internal and force the caller to use `Close`?

File: pkg/test/framework/components/pilot/pilot.go [hunk 3]
```diff
@@ -175,9 +175,22 @@ func NewLocalPilot(namespace string) (LocalPilot, error) {
...
+	ports := ":" + strconv.Itoa(pilotAdsPort)
+	optiopns := &kube.PodSelectOptions{
+		PodNamespace: namespace,
+		PodName:      pod,
+	}
+	forwarder, err := kube.NewPortForwarder(rc, config, optiopns, ports, stopCh)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create port forwarder: %v", err)
+	}
+	if err := forwarder.ForwardPorts(); err != nil {
```
[the remark this answers is not in the document]
Comment (lichuqiang) [reply pkg/test/framework/components/pilot/pilot.go]: I think we have

File: pkg/test/framework/components/policybackend/policybackend.go (hunk 1 above)
Comment (lichuqiang) [reply pkg/test/framework/components/policybackend/policybackend.go]: I meant to, but I can't decide where to place the common func, any suggestions?

File: pkg/test/kube/forwarder.go (hunk 2 above)
Comment (lichuqiang) [reply pkg/test/kube/forwarder.go]: Not really. the Close func here will close all of the opening listeners for portforward,and then lead to the forward go routine to return.
And if the stopCh is closed/triggered, the ForwardPorts() process will return, and entire streamConnection will be destroyed.
There will be no orphan portforward process for both of the two scenarioes. The stopCh is mainly for the scenario of unexpected panic or terminate

Comment (lichuqiang) [reply pkg/test/kube/forwarder.go]: Oh, forget that, I made the thing complex.
So, yeah, we don't need to expose the `stopCh` to end users

**Comment by lichuqiang** [comment]
@nmittler Updated to address the comments (except for the rest client issue). PTAL :)

**Comment by ozevren** [comment]
Thanks for doing this.

/lgtm
There seems to be some conflicts though.

**Comment by lichuqiang** [comment]
@nmittler @ozevren conflicts resolved. PTAL

**Comment by lichuqiang** [comment]
/assign @costinm @yutongz
for changes under test path

**Pull request merged**